evaluator/package_cache.go

```diff
@@ -0,0 +1,139 @@
+package evaluator
+
+import (
+	"app/base/database"
+	"app/base/utils"
+	"sync"
+)
+
+var memoryPackageCache *PackageCache
+
+type PackageCacheMetadata struct {
+	ID              int
+	NameID          int
+	Name            string
+	Evra            string
+	SummaryHash     []byte
+	DescriptionHash []byte
+}
+
+type PackageCache struct {
+	mtx          sync.RWMutex
+	data         []PackageCacheMetadata
+	byID         map[int]*PackageCacheMetadata
+	byNevra      map[string]*PackageCacheMetadata
+	latestByName map[string]*PackageCacheMetadata
+	nameByID     map[int]string
+}
+
+func NewPackageCache() *PackageCache {
+	c := new(PackageCache)
+	c.data = []PackageCacheMetadata{}
+	c.byID = map[int]*PackageCacheMetadata{}
+	c.byNevra = map[string]*PackageCacheMetadata{}
+	c.latestByName = map[string]*PackageCacheMetadata{}
+	c.nameByID = map[int]string{}
+	return c
+}
+
+func (c *PackageCache) load() {
+	tx := database.Db.Begin()
+	defer tx.Rollback()
+	rows, err := tx.Table("package p").
+		Select("p.id, p.name_id, pn.name, p.evra, p.summary_hash, p.description_hash").
+		Joins("JOIN package_name pn ON pn.id = p.name_id").
+		Order("evra DESC").
+		Rows()
+	if err != nil {
+		panic(err)
+	}
+
+	c.mtx = sync.RWMutex{}
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+
+	var columns PackageCacheMetadata
+	for rows.Next() {
+		err = tx.ScanRows(rows, &columns)
+		if err != nil {
+			panic(err)
+		}
+		pkg := PackageCacheMetadata{
+			ID:              columns.ID,
+			NameID:          columns.NameID,
+			Name:            columns.Name,
+			Evra:            columns.Evra,
+			DescriptionHash: columns.DescriptionHash,
+			SummaryHash:     columns.SummaryHash,
+		}
+		c.AddWithoutLock(&pkg)
+	}
+	utils.Log("loaded rows", len(c.data)).Debug("PackageCache.load")
```

Comment:
Good. Don't you want to log [cache size](https://stackoverflow.com/a/26975872) and loading duration? It could be useful.

Reply:
added size and time and changed to info